designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,156 @@
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it.
+
+
+proposed build process in this RFC:
+```
+                                  ┌-------------------┐
+                                  │                   │
+                                  │  dist/espree.cjs  │
+                                  │ (CJS entry point) │                package.json:
+┌-------------------┐             │                   │
+│                   │             └--▲----------------┘     CJS ---▶   "main": "dist/espress.cjs",
+│     espree.js     │                │                                 "exports": {
+│ (ESM entry point) ├───BUILD_STEP───┘                      ESM ---▶     "import": "espree.js",
+│                   │                                       CJS ---▶     "require": "./dist/espree.cjs"
+└-------------------┘                                                  },
+                                                                       "type": "module"
+```
+
+browserify is specifically oriented around commonjs as the input format, so I also propose replacing it with rollup, which understands and expects ESM as it's input format.
+
+`UMD` is not part of the formal package interface and it's usage is questionable, so I propose we drop it altogether.
+
+
+## Documentation
+
+The changes should be described in the migration guide for whichever major version of espree this goes into (currently considering the upcoming 8.x line.)
+
+Adding an `exports` field to `package.json` will break the existing API, it makes sense to formally announce this via blog post.
+
+
+## Drawbacks
+
+The javascript build/tooling ecosystem is already monstrously complex, and this change adds another output format, which further complicates things (compare the 2 graphs above to see this additional complexity visualized.)
+
+I do believe this is a temporary scenario; as time goes on, the UMD and CJS entry points could be dropped altogether, along with the build step. But that will take months, maybe even years depending on how comfortable the community is with this change and how widely adopted the ESM format becomes.
+
+
+## Backwards Compatibility Analysis
+
+The dual package approach _should_ not break espree for anyone currently using espree.
+
+
+## Alternatives
+
+Another option is to  just drop CJS, UMD and the build system altogether, provide an ESM only implementation,
+and make this a part of a semver-major version bump. (e.g., espree@8 is ESM only.)
+
+This would essentially eliminate the build step and make things very simple, but it has a big downside:
+* users that don't support ESM only projects would be stuck on `espree <= 7.x` forever (or at least until they added the ability to use ESM in their own stuff.)
+
```

Comment:
it seems not a big deal:
+ node v10 is going to be EOL in 2021-04-30, then we can drop node <10, and rewrite the codebase to native es module.

Reply:
Please don't do that; converting things to ESM means they can't be `require`d at all. (note, you'd also have to drop node < 12.17, and < 13.7).

The best practice remains writing everything in CJS for the foreseeable future, and writing thin ESM wrappers as needed to provide named exports.

Reply:
@aladdin-add given how widely depending upon eslint stuff is, people here tend to be (rightly so) more conservative in adopting these big changes. I think what you're describing is a totally valid approach for new modules, and it's definitely the strategy I'm using in my own personal module work.

Reply:
@ljharb we can provide a compiled version (esm to cjs using something like babel), and I opened an issue about dropping node versions: https://github.com/eslint/eslint/issues/14023

this helps eslint developers writing esm, rahter than continuing using cjs - cjs will go away finally. :)

Reply:
CJS will never go away.

Certainly you can author in either format and transpile to the other, but i fail to see the advantage of that over just authoring CJS and ESM wrappers without a build step. CJS is precisely as tree-shakeable as ESM, and for node 12 there’s no syntactic advantage either.

Reply:
> CJS will never go away.

Yeah, and 640k of memory ought to be enough for anybody. ;)

Reply:
When we discussed this topic in [a recent TSC meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2020/2020-12-03.md#would-you-be-open-to-a-pure-es-module), we decided to go with a dual package approach to allow a gradual migration. On an infinite time scale we’ll likely drop the build step and go ESM-only, but I wouldn’t want to speculate on a timeline for that.